iXBRLViewerPlugin/xhtmlserialize.py

```diff
@@ -13,29 +13,125 @@
-    selfClosableElements = (
+    SELF_CLOSABLE = (
         'area', 'base', 'basefont', 'br', 'col', 'frame', 'hr', 'img', 
         'input', 'isindex', 'link', 'meta', 'param'
     )
 
-    def _expandEmptyTags(self, xml):
-        """
-        Expand self-closing tags
-
-        Self-closing tags cause problems for XHTML documents when treated as
-        HTML.  Tags that are required to be empty (e.g. <br>) are left as
-        self-closing.
-        """
-        for e in xml.iter('*'):
-            m = re.match(r'\{http://www\.w3\.org/1999/xhtml\}(.*)', e.tag)
-            if (m is None or m.group(1) not in XHTMLSerializer.selfClosableElements) and e.text is None:
-                e.text = ''
-
-    def serialize(self, xmlDocument, fout):
-        self._expandEmptyTags(xmlDocument)
-        fout.write(etree.tostring(xmlDocument, method="xml", encoding="utf-8", xml_declaration=True))
+    ENTITIES = {
+        '>': '&gt;',
+        '<': '&lt;',
+        '"': '&quot;',
+        '&': '&amp;'
+        }
+
+    def __init__(self, xml_declaration = True, assume_xhtml = True):
+        self.xml_declaration = xml_declaration
+        self.assume_xhtml = assume_xhtml
+
+    def prefix_sort(self, p):
+        return p if p is not None else '0'
+
+    def tag_as_qname(self, tag, nsmap):
+        qname = etree.QName(tag)
+        if qname.namespace is None:
+            return qname.localname
+
+        prefix = next(iter(sorted((p for p, ns in nsmap.items() if ns == qname.namespace), key = self.prefix_sort)))
+
+        if prefix is None:
+            return qname.localname
+        else:
+            return "%s:%s" % (prefix, qname.localname)
+
+    def is_selfclosable(self, n):
+        qname = etree.QName(n)
+        return (
+                qname.localname in XHTMLSerializer.SELF_CLOSABLE
+                and (
+                    qname.namespace == XHTML_NS
+                    or (self.assume_xhtml and qname.namespace is None)))
+
+    def escape_attr(self, s):
+        return re.sub(r'([<>"&])', lambda m: self.ENTITIES[m.group(0)], s)
+
+    def xmlns_declaration(self, prefix, uri):
+        if prefix is None:
+            return 'xmlns="%s"' % self.escape_attr(uri)
+        return 'xmlns:%s="%s"' % (prefix, self.escape_attr(uri))
+
+    def namespace_declarations(self, new_nsmap, cur_nsmap):
+        changed = set(p[0] for p in (set(new_nsmap.items()) ^ set(cur_nsmap.items())))
+        return sorted(self.xmlns_declaration(p, new_nsmap[p]) for p in changed)
+
+    def escape_text(self, s, escape_mode):
+        if s is None:
+            return ''
+        if escape_mode == EscapeMode.STYLE:
+            # Don't escape >
+            # & and < may only appear in a string or comment in CSS, so escape
+            # as if they're in a string.
+            return re.sub(r'([<&])', lambda m: "\\%06X" % ord(m.group(1)), s)
+        return re.sub(r'([<>&])', lambda m: self.ENTITIES[m.group(0)], s)
```

Comment:
I guess escaping `>` is never required?  You might add a comment to the "Don't escape" above.

Reply:
Comment expanded.